### Import Package

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

### Import Data 'data_auto.csv'

In [3]:
df = pd.read_csv("Maternal Health Risk Data Set.csv")
df.head()

Age  SystolicBP  DiastolicBP    BS  BodyTemp  HeartRate  RiskLevel
0   25         130           80  15.0      98.0         86  high risk
1   35         140           90  13.0      98.0         70  high risk
2   29          90           70   8.0     100.0         80  high risk
3   30         140           85   7.0      98.0         70  high risk
4   35         120           60   6.1      98.0         76   low risk

### Melihat jumlah baris dan kolom

In [4]:
df.shape

(1014, 7)

terdapat 1014 baris dan 7 kolom<br>
terdapat 6 kolom fitur dan 1 kolom target 

### Melihat data yang kosong

In [4]:
df.replace("?", np.nan, inplace = True)
df.head()

vendor  myct  mmin   mmax  cach  chmin  chmax  prp
0  adviser   125   256   6000   256     16    128  198
1   amdahl    29  8000  32000    32      8     32  269
2   amdahl    29  8000  32000    32      8     32  220
3   amdahl    29  8000  32000    32      8     32  172
4   amdahl    29  8000  16000    32      8     16  132

In [5]:
df.isnull().any()

vendor    False
myct      False
mmin      False
mmax      False
cach      False
chmin     False
chmax     False
prp       False
dtype: bool

In [6]:
df.isnull().sum()

vendor    0
myct      0
mmin      0
mmax      0
cach      0
chmin     0
chmax     0
prp       0
dtype: int64

tidak ada yang missing value atau data yang kosong

### Melihat tipe data

In [8]:
df.dtypes

vendor    object
myct       int64
mmin       int64
mmax       int64
cach       int64
chmin      int64
chmax      int64
prp        int64
dtype: object

### Merubah tipe data sesuai dengan kolom dan deskripsi

In [9]:
df.head()

vendor  myct  mmin   mmax  cach  chmin  chmax  prp
0  adviser   125   256   6000   256     16    128  198
1   amdahl    29  8000  32000    32      8     32  269
2   amdahl    29  8000  32000    32      8     32  220
3   amdahl    29  8000  32000    32      8     32  172
4   amdahl    29  8000  16000    32      8     16  132

In [10]:
df.dtypes

vendor    object
myct       int64
mmin       int64
mmax       int64
cach       int64
chmin      int64
chmax      int64
prp        int64
dtype: object

tipe data pada setiap kolom sudah sesuai dengan deskripsi

### Melihat deskripsi setiap kolom

In [11]:
df.describe()

myct          mmin          mmax        cach       chmin  \
count   209.000000    209.000000    209.000000  209.000000  209.000000   
mean    203.822967   2867.980861  11796.153110   25.205742    4.698565   
std     260.262926   3878.742758  11726.564377   40.628722    6.816274   
min      17.000000     64.000000     64.000000    0.000000    0.000000   
25%      50.000000    768.000000   4000.000000    0.000000    1.000000   
50%     110.000000   2000.000000   8000.000000    8.000000    2.000000   
75%     225.000000   4000.000000  16000.000000   32.000000    6.000000   
max    1500.000000  32000.000000  64000.000000  256.000000   52.000000   

            chmax          prp  
count  209.000000   209.000000  
mean    18.267943   105.622010  
std     25.997318   160.830733  
min      0.000000     6.000000  
25%      5.000000    27.000000  
50%      8.000000    50.000000  
75%     24.000000   113.000000  
max    176.000000  1150.000000

sudah tidak terdapat data yang kosong pada masing-masing kolom

### Melakukan Label Encoding pada kolom kategorikal

In [12]:
df.head()

vendor  myct  mmin   mmax  cach  chmin  chmax  prp
0  adviser   125   256   6000   256     16    128  198
1   amdahl    29  8000  32000    32      8     32  269
2   amdahl    29  8000  32000    32      8     32  220
3   amdahl    29  8000  32000    32      8     32  172
4   amdahl    29  8000  16000    32      8     16  132

In [13]:
df['vendor'].value_counts()

ibm             32
nas             19
ncr             13
sperry          13
honeywell       13
siemens         12
cdc              9
amdahl           9
burroughs        8
hp               7
harris           7
dg               7
dec              6
c.r.d            6
ipl              6
magnuson         6
cambex           5
prime            5
formation        5
perkin-elmer     3
gould            3
nixdorf          3
wang             2
bti              2
basf             2
apollo           2
microdata        1
four-phase       1
sratus           1
adviser          1
Name: vendor, dtype: int64

In [14]:
from sklearn.preprocessing import LabelEncoder

encoding = LabelEncoder()
df['vendor'] = encoding.fit_transform(df['vendor'])

df.head()

vendor  myct  mmin   mmax  cach  chmin  chmax  prp
0       0   125   256   6000   256     16    128  198
1       1    29  8000  32000    32      8     32  269
2       1    29  8000  32000    32      8     32  220
3       1    29  8000  32000    32      8     32  172
4       1    29  8000  16000    32      8     16  132

melakukan proses encoding untuk data fitur berupa teks ataupun kategorikal<br>
agar dapat diproses oleh mesin serta memudahkan dalam melakukan labeling pada data kategorikal

### Menyesuaikan tipe data jika tidak sesuai

In [15]:
df.dtypes

vendor    int32
myct      int64
mmin      int64
mmax      int64
cach      int64
chmin     int64
chmax     int64
prp       int64
dtype: object

tipe data sudah sesuai untuk semua kolom

### Melakukan normalisasi menggunakan MinMax Scaler untuk kolom numerik

In [16]:
df.head()

vendor  myct  mmin   mmax  cach  chmin  chmax  prp
0       0   125   256   6000   256     16    128  198
1       1    29  8000  32000    32      8     32  269
2       1    29  8000  32000    32      8     32  220
3       1    29  8000  32000    32      8     32  172
4       1    29  8000  16000    32      8     16  132

In [17]:
from sklearn.preprocessing import MinMaxScaler
numeric = ["myct", "mmin", "mmax", "cach", "chmin", "chmax"]
scaler = MinMaxScaler(feature_range=(0, 1))

scaler.fit(df.loc[:, df[numeric].columns])
df.loc[:, df[numeric].columns] = scaler.transform(df.loc[:, df[numeric].columns])
df.head()

vendor      myct      mmin      mmax   cach     chmin     chmax  prp
0       0  0.072825  0.006012  0.092843  1.000  0.307692  0.727273  198
1       1  0.008092  0.248497  0.499499  0.125  0.153846  0.181818  269
2       1  0.008092  0.248497  0.499499  0.125  0.153846  0.181818  220
3       1  0.008092  0.248497  0.499499  0.125  0.153846  0.181818  172
4       1  0.008092  0.248497  0.249249  0.125  0.153846  0.090909  132

melakukan normalisasi minmax untuk semua kolom fitur numerikal 

### Dataset Splitting

In [18]:
X = df.drop(columns=["prp"])
y = df["prp"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

tidak memakai stratify karena bukan merupakan klasifikasi dan kolom target berupa numerikal<br>
data training 80%, data testing 20%

### Fitting Dataset menggunakan Linear Regression

In [19]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

print('Training Score\t= ', model.score(X_train, y_train), '\nTesting Score\t= ', model.score(X_test, y_test))

Training Score	=  0.8619691625301994 
Testing Score	=  0.880600143967276


### Melihat nilai Evaluasi data

In [20]:
hasil = model.predict(X_test)
from sklearn.metrics import mean_squared_error, r2_score

print('MSE\t= ', mean_squared_error(y_test, hasil), '\nR2\t= ', r2_score(y_test, hasil))

MSE	=  1843.6710463674997 
R2	=  0.880600143967276
